lib/elixir/lib/calendar/iso.ex

```diff
@@ -953,10 +1007,12 @@ defmodule Calendar.ISO do
   """
   @doc since: "1.5.0"
   @impl true
-  @spec valid_date?(year, month, day) :: boolean
-  def valid_date?(year, month, day) do
-    month in 1..12 and year in -9999..9999 and
-      (is_integer(day) and day >= 1 and day <= days_in_month(year, month))
+  @spec valid_date?(year :: integer, month :: integer, day :: integer) :: boolean
+  def valid_date?(year, month, day)
+      when is_integer(year) and is_integer(month) and is_integer(day) do
+    year >= -9999 and year <= 9999 and
+      month >= 1 and month <= 12 and
+      day >= 1 and day <= days_in_month(year, month)
   end
```

Comment:
I bent the rules here and changed the specs, but if we want to maintain them, this function should be rewritten as :

```elixir
  @spec valid_date?(year, month, day) :: boolean
  def valid_date?(year, month, day) when is_year(year) and is_month(month) and is_day(day) do
    day <= days_in_month(year, month)
  end
```

tests should be changed, since `Calendar.ISO.valid_date?(-1, 12, 32)` would raise `FunctionClauseError`, but I still think it is the way to go since an int out of the range specified by the specs was provided. Let me know if I should change it

Reply:
Yeah, I think this is a good change. Good call!

Reply:
> Yeah, I think this is a good change. Good call!

@josevalim are you refering to my initial code change, or to my later comment https://github.com/elixir-lang/elixir/pull/9493/files#r342723097 ? I was a bit confused by this as I thought you were answering to my comment

Reply:
Your original comment! It is still good as is, thank you!